File: scitbx/math/boost_python/principal_axes_of_inertia.cpp
#include <scitbx/array_family/boost_python/flex_fwd.h>

#include <boost/python/class.hpp>
#include <boost/python/args.hpp>
#include <boost/python/return_value_policy.hpp>
#include <boost/python/copy_const_reference.hpp>
#include <boost/python/return_internal_reference.hpp>
#include <scitbx/math/principal_axes_of_inertia.h>

namespace scitbx { namespace math {
namespace {

  struct principal_axes_of_inertia_wrappers
  {
    typedef principal_axes_of_inertia<> w_t;
    typedef principal_axes_of_inertia_2d<> w_t_2d;

    static void
    wrap()
    {
      using namespace boost::python;
      typedef return_value_policy<copy_const_reference> ccr;
      typedef return_internal_reference<> rir;
      class_<w_t>("principal_axes_of_inertia", no_init)
        .def(init<af::const_ref<vec3<double> > const&>((arg_("points"))))
        .def(init<af::const_ref<vec3<double> > const&,
                  af::const_ref<double> const&>(
          (arg_("points"), arg_("weights"))))
        .def("center_of_mass", &w_t::center_of_mass, ccr())
        .def("inertia_tensor", &w_t::inertia_tensor, ccr())
        .def("eigensystem", &w_t::eigensystem, rir())
        .def("distance_to_inertia_ellipsoid_surface",
          &w_t::distance_to_inertia_ellipsoid_surface, (
            arg_("unit_direction")))
      ;
      class_<w_t_2d>("principal_axes_of_inertia_2d", no_init)
        .def(init<af::const_ref<vec2<double> > const&>((arg_("points"))))
        .def(init<af::const_ref<vec2<double> > const&,
                  af::const_ref<double> const&>(
          (arg_("points"), arg_("weights"))))
        .def("center_of_mass", &w_t_2d::center_of_mass, ccr())
        .def("inertia_tensor", &w_t_2d::inertia_tensor, ccr())
        .def("eigensystem", &w_t_2d::eigensystem, rir())
        .def("distance_to_inertia_ellipsoid_surface",
          &w_t_2d::distance_to_inertia_ellipsoid_surface, (
            arg_("unit_direction")))
      ;
    }
  };

} // namespace <anonymous>

namespace boost_python {

  void wrap_principal_axes_of_inertia()
  {
    principal_axes_of_inertia_wrappers::wrap();
  }

}}} // namespace scitbx::math::boost_python

File: scitbx/examples/principal_axes_of_inertia.py
from __future__ import absolute_import, division, print_function
from scitbx.math import principal_axes_of_inertia
from scitbx.array_family import flex

def run():
  points = flex.vec3_double([
    ( 8.292,  1.817,  6.147),
    ( 9.159,  2.144,  7.299),
    (10.603,  2.331,  6.885),
    (11.041,  1.811,  5.855),
    ( 9.061,  1.065,  8.369),
    ( 7.665,  0.929,  8.902),
    ( 6.771,  0.021,  8.327),
    ( 7.210,  1.756,  9.920),
    ( 5.480, -0.094,  8.796),
    ( 5.904,  1.649, 10.416),
    ( 5.047,  0.729,  9.831),
    ( 3.766,  0.589, 10.291),
    (11.358,  2.999,  7.612)])
  pai = principal_axes_of_inertia(points=points)
  print(pai.center_of_mass())
  print(pai.inertia_tensor())
  es = pai.eigensystem()
  print(list(es.values()))
  print(list(es.vectors()))

if (__name__ == "__main__"):
  run()
